File: src/main/java/com/dafttech/terra/game/world/subtiles/SubtileWater.java
package com.dafttech.terra.game.world.subtiles;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.engine.input.InputHandler$;
import com.dafttech.terra.game.world.Facing;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.resources.Resources$;

public class SubtileWater extends SubtileFluid {
    float img = 0;
    boolean wavephase = false;

    public SubtileWater() {
        super();
    }

    @Override
    public TextureRegion getImage() {
        return Resources$.MODULE$.TILES().getImage("wateranim", (int) img);
    }

    @Override
    public SubtileFluid getNewFluid() {
        return new SubtileWater();
    }

    @Override
    public void onTick(World world, float delta) {
        img += delta;
        if ((int) img > 3) img = 0;

        float windSpeed = world.weather().getWindSpeed(world);
        Facing facing = windSpeed > 0 ? Facing.RIGHT$.MODULE$ : windSpeed < 0 ? Facing.LEFT$.MODULE$ : null;
        if (InputHandler$.MODULE$.isKeyDown("WAVESLEFT")) {
            facing = Facing.LEFT$.MODULE$;
            windSpeed = 5;
        } else if (InputHandler$.MODULE$.isKeyDown("WAVESRIGHT")) {
            facing = Facing.RIGHT$.MODULE$;
            windSpeed = 5;
        }
        if (facing != null && !wavephase) {
            wavephase = true;
            SubtileFluid fluid = getFluid(world, facing);
            int maxReach = 5;
            float amount = maxPressure * delta * Math.abs(windSpeed);
            while (maxReach > 0 && fluid != null && pressure > amount) {
                maxReach--;
                if (fluid.pressure + amount < fluid.maxPressure * 2) {
                    addPressure(-amount);
                    fluid.addPressure(amount);
                }
                fluid = fluid.getFluid(world, facing);
            }
        } else {
            wavephase = false;
            super.onTick(world, delta);
        }
    }

    @Override
    public float getViscosity() {
        return (InputHandler$.MODULE$.isKeyDown("WAVESRIGHT") || InputHandler$.MODULE$.isKeyDown("WAVESLEFT")) ? 4 : 0;
    }

    @Override
    public int getMaxReach() {
        return (InputHandler$.MODULE$.isKeyDown("WAVESRIGHT") || InputHandler$.MODULE$.isKeyDown("WAVESLEFT")) ? 4 : 10;
    }

    @Override
    public int getPressCap() {
        return 2;
    }

    @Override
    public boolean providesSunlightFilter() {
        return true;
    }

    @Override
    public Color getFilterColor() {
        return new Color(1 - pressure / maxPressure, 1 - pressure / maxPressure, 1, 1);
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/subtiles/SubtileWater.scala
package com.dafttech.terra.game.world.subtiles

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.engine.input.InputHandler
import com.dafttech.terra.game.world.{Facing, World}
import com.dafttech.terra.resources.Resources

class SubtileWater() extends SubtileFluid {
  private[subtiles] var img: Float = 0
  private[subtiles] var wavephase = false

  override def getImage: TextureRegion = Resources.TILES.getImage("wateranim", img.toInt)

  override def getNewFluid = new SubtileWater

  override def onTick(delta: Float)(implicit tilePosition: TilePosition): Unit = {
    img += delta
    if (img.toInt > 3) img = 0
    var windSpeed = tilePosition.world.weather.getWindSpeed(tilePosition.world)
    var facing = if (windSpeed > 0) Facing.Right
    else if (windSpeed < 0) Facing.Left
    else null
    if (InputHandler.isKeyDown("WAVESLEFT")) {
      facing = Facing.Left
      windSpeed = 5
    }
    else if (InputHandler.isKeyDown("WAVESRIGHT")) {
      facing = Facing.Right
      windSpeed = 5
    }
    if (facing != null && !wavephase) {
      wavephase = true
      var fluid = getFluid(tilePosition, facing)
      var maxReach = 5
      val amount = maxPressure * delta * Math.abs(windSpeed)
      while ( {
        maxReach > 0 && fluid != null && pressure > amount
      }) {
        maxReach -= 1
        if (fluid._1.pressure + amount < fluid._1.maxPressure * 2) {
          addPressure(-amount)
          fluid._1.addPressure(amount)
        }
        fluid = fluid._1.getFluid(tilePosition, facing)
      }
    }
    else {
      wavephase = false
      super.onTick(delta)
    }
  }

  override def getViscosity: Float = if (InputHandler.isKeyDown("WAVESRIGHT") || InputHandler.isKeyDown("WAVESLEFT")) 4 else 0

  override def getMaxReach: Int = if (InputHandler.isKeyDown("WAVESRIGHT") || InputHandler.isKeyDown("WAVESLEFT")) 4 else 10

  override def getPressCap: Int = 2

  override def providesSunlightFilter: Boolean = true

  override def getFilterColor = new Color(1 - pressure / maxPressure, 1 - pressure / maxPressure, 1, 1)
}